### DISCUSSION TITLE
host cloud anchor only after anchor is placed.

### DISCUSSION THREAD

**Discussion by tony-covais** [sceneview/sceneform-android#352] [Q&A] [answered]
Hello, very new to android development and development in general. I'm following a tutorial on placing objects and hosting cloud anchors; I've got everything working but I'd like to modify things a bit. How can I place a single model in a scene (currently just using shapes from the ModelFactory) and then only host the anchor once the user has placed the object where and how they'd like it?

The onTapListener code I'm currently running:

``` Java
        MaterialFactory.makeOpaqueWithColor(this, new Color(android.graphics.Color.BLUE))
                .thenAccept(material -> {
                    sphereRenderable = ShapeFactory.makeSphere(0.1f, new Vector3(0.0f,0.15f, 0.0f ),material);
                });

        arFragment = (CustomArFragment) getSupportFragmentManager().findFragmentById(R.id.ux_fragment);
        assert arFragment != null;
        arFragment.getArSceneView().getScene().addOnUpdateListener(this::onUpdateFrame);


        assert arFragment != null;
        arFragment.setOnTapArPlaneListener(
                (HitResult hitResult, Plane plane, MotionEvent motionEvent) -> {
                    if (plane.getType() != Plane.Type.HORIZONTAL_UPWARD_FACING ||
                            appAnchorState != AppAnchorState.NONE) {
                        return;
                    }
                    Anchor anchor = arFragment.getArSceneView().getSession().hostCloudAnchor(hitResult.createAnchor());

                    setCloudAnchor(anchor);

                    appAnchorState = AppAnchorState.HOSTING;
                    Toast.makeText(this,"Now Hosting...", Toast.LENGTH_LONG).show();
                    snackbarHelper.showMessage(this, "Now hosting...");

                    AnchorNode anchorNode = new AnchorNode(anchor);
                    anchorNode.setParent(arFragment.getArSceneView().getScene());

                    TransformableNode sphere = new TransformableNode(arFragment.getTransformationSystem());
                    sphere.setParent(anchorNode);
                    sphere.setRenderable(sphereRenderable);
                    isPlaced = true;
                    sphere.select();

                }
        );
```

I believe I would first have to set the initial anchor to hitResult.createAnchor(); but how would I allow the anchor to be hosted only after it is placed in a scene? This code begins hosting the anchor immediately. I planned to have a button revealed once the object is placed called "Finish" which would host the anchor. Thank you for any help.

**Comment by grassydragon** [comment]
Hi!
It is recommended to first make sure that the feature map quality is sufficient before hosting the Cloud Anchor: https://developers.google.com/ar/develop/java/cloud-anchors/developer-guide#check_the_mapping_quality_of_feature_points
For example, this ARCore sample application has a UI that helps the user to scan the anchor from different points of view (you don't need to set up Firebase to see the UI): https://github.com/google-ar/arcore-android-sdk/tree/master/samples/persistent_cloud_anchor_java

To allow the user to host the Cloud Anchor after he or she confirms the object position you can simply use a button that becomes visible or enabled after the object is placed.

**Reply by tony-covais** [reply]
I've tried to set the anchor to a global variable and then:


``` Java
        anchor = hitResult.createAnchor();

        AnchorNode anchorNode = new AnchorNode(anchor);

        anchorNode.setParent(arFragment.getArSceneView().getScene());

        TransformableNode sphere = new TransformableNode(arFragment.getTransformationSystem());

        sphere.setParent(anchorNode);

        sphere.setRenderable(sphereRenderable);

        isPlaced = true;

        sphere.select();
```

inside of the the onTapListener. Inside of my Finish button onClick listener, I try:

``` Java
                Session session = arFragment.getArSceneView().getSession();

                session.hostCloudAnchor(anchor);

                appAnchorState = AppAnchorState.HOSTING;

                Toast.makeText(getApplicationContext(),"Now Hosting...", Toast.LENGTH_LONG).show();

                checkUpdatedAnchor();
```

I have a onFrameUpdateListener active, which should begin to store the anchor ID into a database; but the anchor doesn't seem to be hosting. Any suggestions?

**Comment by tony-covais** [comment, marked as answer]
Answered! Thanks, @grassydragon . The object that is returned by the hostCloudAnchor() method wasn't being used in the code previously.
